spatialdata/_core/aggregate.py

```diff
@@ -0,0 +1,164 @@
+    """
+    # TODO: Check that values are in the same space
+    # Dispatch
+    by_type = get_schema(by)
+    values_type = get_schema(values)
+    if by_type is ShapesModel:
+        if values_type is PointsModel:
+            return _aggregate_points_by_shapes(values, by, id_key, value_key=value_key, agg_func=agg_func)
+        elif values_type is ShapesModel:
+            return _aggregate_shapes_by_shapes(values, by, id_key, value_key=value_key, agg_func=agg_func)
+        else:
+            raise NotImplementedError(f"Cannot aggregate {values_type} by {by_type}")
+    else:
+        raise NotImplementedError(f"Cannot aggregate {values_type} by {by_type}")
+
+
+def _aggregate_points_by_shapes(
+    points: ddf.DataFrame | pd.DataFrame,
+    shapes: gpd.GeoDataFrame,
+    id_key: str,
+    *,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    # Have to get dims on dask dataframe, can't get from pandas
+    dims = sd.get_dims(points)
+    if isinstance(points, ddf.DataFrame):
+        points = points.compute()
+    points = gpd.GeoDataFrame(points, geometry=gpd.points_from_xy(*[points[dim] for dim in dims]))
+    return _aggregate(points, shapes, id_key, value_key, agg_func)
+
+
+def _aggregate_shapes_by_shapes(
+    values: gpd.GeoDataFrame,
+    by: gpd.GeoDataFrame,
+    id_key: str,
+    *,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    def circles_to_polygons(df: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+        # We should only be buffering points, not polygons. Unfortunately this is an expensive check.
+        values_geotypes = list(values.geom_type.unique())
+        if values_geotypes == ["Point"]:
+            df = df.set_geometry(df.geometry.buffer(df[ShapesModel.RADIUS_KEY]))
+        elif "Point" in values_geotypes:
+            raise TypeError("Geometry contained shapes and polygons.")
+        return df
+
+    values = circles_to_polygons(values)
+    by = circles_to_polygons(by)
+
+    return _aggregate(values, by, id_key, value_key, agg_func)
+
+
+def _aggregate(
+    value: gpd.GeoDataFrame,
+    by: gpd.GeoDataFrame,
+    id_key: str,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    """
+    Inner function to aggregate geopandas objects.
+
+    See docstring for `aggregate` for semantics.
+
+    Parameters
+    ----------
+    value
+        Geopandas dataframe to be aggregated. Must have a geometry column.
+    by
+        Geopandas dataframe to group values by. Must have a geometry column.
+    id_key
+        Column in value dataframe to group values by.
+    value_key
+        Column in value dataframe to perform aggregation on.
+    agg_func
+        Aggregation functio to apply over grouped values. Passed to pandas.DataFrame.groupby.agg.
+    """
+    assert pd.api.types.is_categorical_dtype(value[id_key]), f"{id_key} must be categorical"
+
+    if by.index.name is None:
+        by.index.name = "cell"
```

Comment:
not sure `"cell"` is a good option here. E.g. in the visium/xenium example they are not cells but spot. I'd rather not use this info at all, or use a generic "index" (and still maybe not hardcode it but allowing user to pass it).

Reply:
Could do `"index"`. Open to other names as well. Note that this only is used if the index doesn't already have a name, so I think that would cover being "user passed".

Reply:
I'm actually going to ignore these, since they are changes to private docs that won't be rendered.